iceoryx_posh/source/popo/ports/server_port_roudi.cpp

```diff
@@ -38,19 +39,108 @@ ServerPortRouDi::MemberType_t* ServerPortRouDi::getMembers() noexcept
     return reinterpret_cast<MemberType_t*>(BasePort::getMembers());
 }
 
+ConsumerTooSlowPolicy ServerPortRouDi::getClientTooSlowPolicy() const noexcept
+{
+    return static_cast<ConsumerTooSlowPolicy>(getMembers()->m_chunkSenderData.m_subscriberTooSlowPolicy);
+}
+
 cxx::optional<capro::CaproMessage> ServerPortRouDi::tryGetCaProMessage() noexcept
 {
+    // get offer state request from user side
+    const auto offeringRequested = getMembers()->m_offeringRequested.load(std::memory_order_relaxed);
+
+    const auto isOffered = getMembers()->m_offered.load(std::memory_order_relaxed);
+
+    if (isOffered)
+    {
+        if (!offeringRequested)
+        {
+            capro::CaproMessage caproMessage(capro::CaproMessageType::STOP_OFFER, this->getCaProServiceDescription());
+            return dispatchCaProMessageAndGetPossibleResponse(caproMessage);
+        }
+    }
+    else
+    {
+        if (offeringRequested)
+        {
+            capro::CaproMessage caproMessage(capro::CaproMessageType::OFFER, this->getCaProServiceDescription());
+            return dispatchCaProMessageAndGetPossibleResponse(caproMessage);
+        }
+    }
+
     // nothing to change
-    return cxx::nullopt_t();
+    return cxx::nullopt;
 }
 
 cxx::optional<capro::CaproMessage>
-ServerPortRouDi::dispatchCaProMessageAndGetPossibleResponse(const capro::CaproMessage& /*caProMessage*/) noexcept
+ServerPortRouDi::dispatchCaProMessageAndGetPossibleResponse(const capro::CaproMessage& caProMessage) noexcept
 {
-    capro::CaproMessage responseMessage(
-        capro::CaproMessageType::NACK, this->getCaProServiceDescription(), capro::CaproMessageSubType::NOSUBTYPE);
+    const auto isOffered = getMembers()->m_offered.load(std::memory_order_relaxed);
+
+    if (isOffered)
```

Comment:
Personal taste: I would prefer something like:
```cpp
return (isOffered) ? handleCapProMessageForStateOffered(caProMessage)
                   : handleCapProMessageForStateNotOffered(caProMessage);
```

So two lines instead of 8.

Reply:
Good point